tools/src/scie_pants/pants_version.py

```diff
@@ -162,5 +182,120 @@ def configure_version():
+                f"It may also be a missing Pants PEX entry in the PANTS_BOOTSTRAP_URLS file {bootstrap_urls_path} for "
+                f"a name like `pants.{version}-cpNN-{platform}.pex`.\n\n"
+            )
+            if bootstrap_urls_path
+            else ""
+        )
+        fatal(
+            f"Unknown Pants release: {version}.\n\n"
+            "If this is unexpected (you are using a known good Pants version), try upgrading scie-pants first.\n"
+            f"It may also be that the platform {platform} isn't supported for this version of Pants.\n"
+            + bootstrap_urls_info
+            + "To get help, please visit: https://www.pantsbuild.org/community/getting-help"
+        )
+    pex_url, python = res
+    if python not in PYTHON_IDS:
+        # Should not happen... but if we mess up, this is a nicer error message rather than blowing up.
+        fatal(f"This version of scie-pants does not support {python!r}.")
+    return pex_url, PYTHON_IDS[python]
+
+
+def get_pex_url_and_python_id(
+    ptex: Ptex,
+    version: Version,
+    platform: str,
+    bootstrap_urls_path: str | None,
+) -> tuple[str, str] | None:
+    if bootstrap_urls_path:
+        bootstrap_urls = json.loads(Path(bootstrap_urls_path).read_text())
+        ptex_urls = bootstrap_urls.get("ptex")
+        if ptex_urls is None:
+            raise ValueError(
+                f"Missing 'ptex' key in PANTS_BOOTSTRAP_URLS file: {bootstrap_urls_path}"
+            )
+    else:
+        ptex_urls = None
+
+    py = get_python_id_for_pants_version(version)
+    if py:
+        pex_url = get_download_url(version, platform, py, ptex_urls)
+        if pex_url:
+            return pex_url, py
+
+    # Else, try all known Pythons...
+    for maybe_py in PYTHON_IDS.keys():
+        pex_url = get_download_url(version, platform, maybe_py, ptex_urls)
+        if pex_url:
+            return pex_url, maybe_py
+
+    return None
+
+
+def get_python_id_for_pants_version(version: Version) -> str | None:
+    versions = json.loads(importlib.resources.read_text("scie_pants", "pants_python_versions.json"))
+    for version_break in versions:
+        # We pick the first version that is less than or equal to the one we're looking for. The
+        # list of pants versions must therefore be sorted in descending order.
+        if Version(version_break["pants"]) <= version:
+            return version_break["python"]
+
+    return None
+
+
+def get_download_url(
+    version: Version, platform: str, python: str, ptex_urls: dict | None
+) -> str | None:
+    pex_name = f"pants.{version}-{python}-{platform}.pex"
+    if ptex_urls:
+        pex_url = ptex_urls.get(pex_name)
+        if not pex_url:
+            return None
+    else:
+        pex_url = (
+            f"https://github.com/pantsbuild/pants/releases/download/release_{version}/{pex_name}"
+        )
+    req = urllib.request.Request(pex_url, method="HEAD")
+    try:
+        with urllib.request.urlopen(req) as rsp:
```

Comment:
I think we should probably set a timeout here, e.g. 
```suggestion
        with urllib.request.urlopen(req, timeout=10) as rsp:
```

Potentially it should be configurable too?

Reply:
Using env var `PANTS_BOOTSTRAP_URL_REQUEST_TIMEOUT_SECONDS`, defaulting to `10`.